### PR TITLE
Update KEYCLOAK_SESSION cookie to not have sessionId in plaintext

### LINKED ISSUE #34026: Update KEYCLOAK_SESSION cookie to not have sessionId in plain-text
### Description

Could we improve the `KEYCLOAK_SESSION` cookie to not contain sessionId in plain-text? This can be considered as possible hardening. Could be nice especially due the fact that `KEYCLOAK_SESSION` cookie is not httpOnly.

Can we possibly hash the sessionId (or maybe whole cookie value) with something like SHA-256 instead of using plain-text value? It looks that this won't require any changes in keycloak.js adapter itself, so it can be backwards compatible for people using older versions of keycloak.js. The keycloak.js can still send the sessionId in the message, which would be consumed by `login-status-iframe.html`. But the iframe itself can then hash the value of the received sessionId and compare it with the hashed value from the cookie.

There are few more places in Keycloak, which may require some changes as they compare the cookie session with the existing session (`LogoutEndpoint,`, `AuthenticationManager.redirectAfterSuccessfulFlow`) , but looks all of them can be updated to compare with the hash instead.

There is also one consideration that `KEYCLOAK_SESSION` cookie now contains the format like `realmId/userId/sessionId` . The "realm" and "user" parts should not be needed for anything, so can be just omitted, so we have the new format of the cookie something like `sha256(sessionId)` (This was already discussed in the past and we might have other open issue for this).

NOTE: This change could be a concern for backwards compatibility for the cluster-like deployments, which would use the new servers (with this change) combined with the old servers (without this change). That is due the format of the cookie would be incompatible. So if we want to support zero-downtime upgrade for cluster deployments, then this change probably should not go to the micro-release, but only to the minor releases.

### Discussion

_No response_

### Motivation

_No response_

### Details

_No response_

### PR DESCRIPTION
**Pull request by mposolda**
closes #34026

- Updated `KEYCLOAK_SESSION` state cookie to not have sessionId in plain-text. More details described in the #34026

- This handles also the backwards compatibility, so the migration from previous versions works as expected. After migration from previous version, user is not logged-out. The verification in `login-status-iframe.html` as well as in `AuthenticationManager.compareSessionIdWithSessionCookie` also takes the old format of the cookie into account.

### CODE REVIEW

File: services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java [hunk 1]
```diff
@@ -960,19 +956,31 @@ public static Response redirectAfterSuccessfulFlow(KeycloakSession session, Real
...
-    public static String getSessionIdFromSessionCookie(KeycloakSession session) {
+    /**
+     * @param session keycloak session
+     * @param sessionId in plain-text
+     * @return true if sessionId matches with the session from KEYCLOAK_SESSION_COOKIE
+     */
+    public static boolean compareSessionIdWithSessionCookie(KeycloakSession session, String sessionId) {
+        if (sessionId == null) {
+            throw new IllegalArgumentException("Not expected to provide null sessionId");
+        }
```
Comment (jonkoops) [inline services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java:967]: Not sure if possible, but why not use the `assert` keyword here? See: https://www.baeldung.com/java-assert

File: services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java [hunk 2]
```diff
@@ -960,19 +956,31 @@ public static Response redirectAfterSuccessfulFlow(KeycloakSession session, Real
...
+            return false;
         }
 
-        String[] parts = cookie.split("/", 3);
-        if (parts.length != 3) {
-            logger.debugv("Cannot parse session value from: {0}", KEYCLOAK_SESSION_COOKIE);
-            return null;
+        if (cookie.equals(sha256UrlEncodedHash(sessionId))) return true;
+
+        // Backwards compatibility
```
Comment (jonkoops) [inline services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java:977]: Should we log an issue to remove this in the next major?

File: services/src/main/resources/org/keycloak/protocol/oidc/endpoints/login-status-iframe.html [hunk 3]
```diff
@@ -72,6 +72,10 @@
 
         // If the client and origin from the event match the verified ones from the server, signal if the cookie has changed.
         if (clientId === init.clientId && origin === init.origin) {
+          const hashedSessionId = await hashString(sessionState);
+          if (hashedSessionId === cookie) return "unchanged";
+
+          // Backwards compatibility with versions older than 26.1
```
Comment (jonkoops) [inline services/src/main/resources/org/keycloak/protocol/oidc/endpoints/login-status-iframe.html:78]: Same for this code, perhaps we log an issue to remove this in the next major.

File: services/src/main/resources/org/keycloak/protocol/oidc/endpoints/login-status-iframe.html [hunk 4]
```diff
@@ -120,6 +124,42 @@
...
+
+      /**
+        * @param {string} message
+        * @see https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/digest#basic_example
+        */
+      async function sha256Digest(message) {
+        const encoder = new TextEncoder();
+        const data = encoder.encode(message);
+
+        if (typeof crypto === "undefined" || typeof crypto.subtle === "undefined") {
```
Comment (jonkoops) [inline services/src/main/resources/org/keycloak/protocol/oidc/endpoints/login-status-iframe.html:145]: Note that the Web Crypto API might not be available in all deployments of Keycloak. We might have to load the `web-crypto-shim.js` on [insecure contexts](https://developer.mozilla.org/en-US/docs/Web/Security/Secure_Contexts) (e.g. a non local IP, or a host without HTTPS).

See:
https://github.com/keycloak/keycloak/blob/8855cf231690e2661f777d8f78214fe9c2826031/js/apps/admin-ui/maven-resources/theme/keycloak.v2/admin/index.ftl#L82-L84

File: services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java (hunk 1 above)
Comment (mposolda) [reply services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java:965]: There is no reason to not use `assert` . It is probably slightly better since it is 1 line instead of 3, so I've just updated to use that.

File: services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java (hunk 2 above)
Comment (mposolda) [reply services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java:975]: Logged https://github.com/keycloak/keycloak/issues/34772

File: services/src/main/resources/org/keycloak/protocol/oidc/endpoints/login-status-iframe.html (hunk 3 above)
Comment (mposolda) [reply services/src/main/resources/org/keycloak/protocol/oidc/endpoints/login-status-iframe.html:81]: Logged https://github.com/keycloak/keycloak/issues/34772

File: services/src/main/java/org/keycloak/protocol/oidc/endpoints/IframeUtil.java [hunk 5]
```diff
@@ -38,7 +44,7 @@ public static Response returnIframeFromResources(String fileName, String version
             cacheControl = CacheControlUtil.noCache();
         }
 
-        InputStream resource = IframeUtil.class.getResourceAsStream(fileName);
+        Object resource = responseEntityProvider.get();
```
Comment (jonkoops) [inline services/src/main/java/org/keycloak/protocol/oidc/endpoints/IframeUtil.java:47]: Is it normal for this to be typed as `Object`? Or is there a more specific type available?

File: services/src/main/java/org/keycloak/protocol/oidc/endpoints/LoginStatusIframeEndpoint.java [hunk 6]
```diff
@@ -51,7 +55,20 @@ public LoginStatusIframeEndpoint(KeycloakSession session) {
     @GET
     @Produces(MediaType.TEXT_HTML_UTF_8)
     public Response getLoginStatusIframe(@QueryParam("version") String version) {
-        return returnIframeFromResources("login-status-iframe.html", version, session);
+        final var map = new HashMap<String, Object>();
+        final var isSecureContext = SecureContextResolver.isSecureContext(session);
+        final var adminBaseUri = session.getContext().getUri(UrlType.ADMIN).getBaseUri();
```
Comment (jonkoops) [inline services/src/main/java/org/keycloak/protocol/oidc/endpoints/LoginStatusIframeEndpoint.java:60]: I wonder if we should be using a different endpoint here. What happens if the admin console is [exposed on a different hostname](https://www.keycloak.org/server/hostname#_exposing_the_administration_console_on_a_separate_hostname) that is not publicly available? Perhaps this should be `UrlType.FRONTEND`? Note I am not fully aware of the implications of this due to lack of JavaDoc on this enum.

File: services/src/main/java/org/keycloak/protocol/oidc/endpoints/LoginStatusIframeEndpoint.java [hunk 7]
```diff
@@ -51,7 +55,20 @@ public LoginStatusIframeEndpoint(KeycloakSession session) {
...
+        map.put("isSecureContext", isSecureContext);
+        map.put("resourceCommonUrl", Urls.themeRoot(adminBaseUri).getPath() + "/common/keycloak");
+
+        return IframeUtil.returnIframe(version, session, () -> {
+            try {
+                return FreemarkerUtils.loadTemplateFromClasspath(map, "login-status-iframe.ftl", getClass());
+            } catch (Exception e) {
+                logger.error("Failure when loading login-status-iframe.ftl", e);
+                return null;
```
Comment (jonkoops) [inline services/src/main/java/org/keycloak/protocol/oidc/endpoints/LoginStatusIframeEndpoint.java:69]: Would this not trigger a null pointer later on since `getResourceAsStream()` will be called with this `null` value? Perhaps it might just as well error out here? Not sure what is considered best practice in this case.

File: services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java [hunk 8]
```diff
@@ -803,20 +804,18 @@ public static IdentityCookieToken createIdentityToken(KeycloakSession keycloakSe
     }
 
     public static void createLoginCookie(KeycloakSession keycloakSession, RealmModel realm, UserModel user, UserSessionModel session, UriInfo uriInfo, ClientConnection connection) {
+        assert session != null : "User session cannot be null";
```
Comment (jonkoops) [inline services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java:807]: I was mistaken about `assert`, it seems that it requires a runtime flag for it to work properly, and is therefore not really a valid option. We could use `org.wildfly.common.Assert` here though:

```java
Assert.checkNotNullParam("session", session);
```

Alternatively, since this is such a common case and we likely don't want to further depend on anything WildFly we just use the same old null check.


```java
if (session == null) {
    throw new IllegalArgumentException("The 'session' parameter cannot be null.");
}
```

File: services/src/main/resources/org/keycloak/protocol/oidc/endpoints/login-status-iframe.ftl [hunk 9]
```diff
@@ -49,7 +52,7 @@
             preventAdditionalRequests = true;
           }
 
-          const url = new URL(`${location.origin}${location.pathname}/init`);
+          const url = new URL(window.location.origin + window.location.pathname + "/init");
```
Comment (jonkoops) [inline services/src/main/resources/org/keycloak/protocol/oidc/endpoints/login-status-iframe.ftl:55]: Let's not change this code

File: services/src/main/java/org/keycloak/protocol/oidc/endpoints/IframeUtil.java (hunk 5 above)
Comment (mposolda) [reply services/src/main/java/org/keycloak/protocol/oidc/endpoints/IframeUtil.java:47]: Yes, you can see that method `jakarta.ws.rs.Resource.ok` takes `Object` as an argument. You can pass various types of objects here like the:
- String (That is what is used in the case of freemarker template)
- InputStream (That is what is used in the case of `IframeUtil.returnIframeFromResources` which is used for loading html files https://github.com/keycloak/keycloak/pull/34551/files#diff-1bad86b3a44fb4631eddecd1d9f7ec38f93a6450a8161ad2babe8163ef59b2afR33 )
- Some more types (You can see Keycloak sources for the inspiration of what we're passing here in various places)

File: services/src/main/java/org/keycloak/protocol/oidc/endpoints/LoginStatusIframeEndpoint.java (hunk 7 above)
Comment (mposolda) [reply services/src/main/java/org/keycloak/protocol/oidc/endpoints/LoginStatusIframeEndpoint.java:69]: No, it does not call `getResourceAsStream()` and it won't trigger `NullPointerException` . When this method returns null, it is handled here on this line https://github.com/keycloak/keycloak/pull/34551/files#diff-1bad86b3a44fb4631eddecd1d9f7ec38f93a6450a8161ad2babe8163ef59b2afR48 and you can see in the following code in `IframeUtil.returnIframe` that it will return `Response.Status.NOT_FOUND` (aka HTTP error 404). This is better than re-throwing the exception as that will cause returning `500`, which should not really be done. In case that file is not found, the `404` is better practice than `500` .

File: services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java (hunk 8 above)
Comment (mposolda) [reply services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java]: Ok, I've replaced it with `Objects.requireNonNull` . Not really big deal as `session` passed to this method is never null currently, so this is more convenient just for the future if someone uses this method in the future.

File: services/src/main/resources/org/keycloak/protocol/oidc/endpoints/login-status-iframe.ftl (hunk 9 above)
Comment (mposolda) [reply services/src/main/resources/org/keycloak/protocol/oidc/endpoints/login-status-iframe.ftl:55]: I've changed this just because freemarker has issues with the `$` notation.

Isn't this effectively the same as before?

File: services/src/main/java/org/keycloak/protocol/oidc/endpoints/IframeUtil.java (hunk 5 above)
Comment (jonkoops) [reply services/src/main/java/org/keycloak/protocol/oidc/endpoints/IframeUtil.java:47]: Thanks for the clarification!

File: services/src/main/java/org/keycloak/protocol/oidc/endpoints/LoginStatusIframeEndpoint.java (hunk 7 above)
Comment (jonkoops) [reply services/src/main/java/org/keycloak/protocol/oidc/endpoints/LoginStatusIframeEndpoint.java:69]: Are you sure? It seems this lambda is [evaluated here](https://github.com/keycloak/keycloak/blob/462613bc328040bec87399698a3aa33e6d7fc431/services/src/main/java/org/keycloak/protocol/oidc/endpoints/IframeUtil.java#L33) where the result is passed directly into `IframeUtil.class.getResourceAsStream()`.

File: services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java [hunk 10]
```diff
@@ -960,19 +954,31 @@ public static Response redirectAfterSuccessfulFlow(KeycloakSession session, Real
...
-    public static String getSessionIdFromSessionCookie(KeycloakSession session) {
+    /**
+     * @param session keycloak session
+     * @param sessionId in plain-text
+     * @return true if sessionId matches with the session from KEYCLOAK_SESSION_COOKIE
+     */
+    public static boolean compareSessionIdWithSessionCookie(KeycloakSession session, String sessionId) {
+        if (sessionId == null) {
+            throw new IllegalArgumentException("Not expected to provide null sessionId");
+        }
```
Comment (jonkoops) [inline services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java:965]: Perhaps for consistency?
```java
Objects.requireNonNull(sessionId, "Session id cannot be null");
```

File: services/src/main/resources/org/keycloak/protocol/oidc/endpoints/login-status-iframe.html (hunk 4 above)
Comment (jonkoops) [reply services/src/main/resources/org/keycloak/protocol/oidc/endpoints/login-status-iframe.html:148]: This is great, I'll resolve this thread 👍

File: services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java (hunk 10 above)
Comment (mposolda) [reply services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java]: +1, updated

File: services/src/main/resources/org/keycloak/protocol/oidc/endpoints/login-status-iframe.ftl (hunk 9 above)
Comment (mposolda) [reply services/src/main/resources/org/keycloak/protocol/oidc/endpoints/login-status-iframe.ftl]: Thanks, updated

File: services/src/main/java/org/keycloak/protocol/oidc/endpoints/LoginStatusIframeEndpoint.java (hunk 7 above)
Comment (mposolda) [reply services/src/main/java/org/keycloak/protocol/oidc/endpoints/LoginStatusIframeEndpoint.java:69]: Yes, it works like this:

When the `LoginStatusIframeEndpoint.getLoginStatusIframe` is called:
- The `LoginStatusIframeEndpoint.getLoginStatusIframe` invokes the `IframeUtil.returnIframe` (not the method `IframeUtil.returnIframeFromResources`)
- When the `LoginStatusIframeEndpoint.getLoginStatusIframe` calls the line `responseEntityProvider.get()`, it invokes the supplier, which was passed to it as an argument. This is the code in `LoginStatusIframeEndpoint` - [lines 65-70](https://github.com/keycloak/keycloak/pull/34551/files#diff-85c707c42bef5061b6a4833b7fdc84ae2083f35372338bbea1535ce9c65a7ffaR65-R69) .
- If this code returned null, it is evaluated on [this line](https://github.com/keycloak/keycloak/pull/34551/files#diff-1bad86b3a44fb4631eddecd1d9f7ec38f93a6450a8161ad2babe8163ef59b2afR48) to return 404
- Note there is not any interaction at all with the method `IframeUtil.returnIframeFromResources` and hence with the `IframeUtil.class.getResourceAsStream`)


When the `ThirdPartyCookiesIframeEndpoint.step1` (or `step2` is called)
- It invokes the `IframeUtil.returnIframeFromResources`
- When the `LoginStatusIframeEndpoint.getLoginStatusIframe` calls the line `responseEntityProvider.get()`, it invokes the supplier, which was passed to it as an argument. This is the code in `LoginStatusIframeEndpoint.getLoginStatusIframe`, which invokes the `IframeUtil.class.getResourceAsStream`
- In other words, this code for handling HTML files works exactly the same as before this PR. There is the supplier, which invokes the `IframeUtil.class.getResourceAsStream()`, which is exactly same behaviour as before.

Is it more clear now?

Comment (jonkoops) [reply services/src/main/java/org/keycloak/protocol/oidc/endpoints/LoginStatusIframeEndpoint.java:69]: Yes, thanks for the explainer!

**Review by jonkoops — APPROVED** [review]
LGTM!

**Comment by mposolda** [comment]
@jonkoops Thanks!
